File: src/leetcode/5.longest-palindromic-substring.js
/*
 * @lc app=leetcode id=5 lang=javascript
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function (s) {
    if (!s) return ''
    if (s.length === 1) return s
    if (s.length === 2) return s[0] === s[1] ? s : s[1]
    let maxStr = '', len = s.length

    for (let i = 0; i < len; i++) {
        let even = '', odd = ''
        if (s[i] === s[i + 1]) {
            let evenIndex = center(s, i - 1, i + 2)
            even = s.slice(evenIndex.left, evenIndex.right)
        }
        let oddIndex = center(s, i - 1, i + 1)
        odd = s.slice(oddIndex.left, oddIndex.right)

        let logger = even.length > odd.length ? even : odd
        maxStr = maxStr.length > logger.length ? maxStr : logger
    }
    return maxStr
}

var center = (s, left, right) => {
    let len = s.length
    while (left >= 0 && right < len && s[left] === s[right]) {
        left--
        right++
    }
    return { left: left + 1, right: right }
}
// @lc code=end

File: src/leetcode/5.longest-palindromic-substring.go
package leetcode

/*
 * @lc app=leetcode id=5 lang=golang
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
func LongestPalindrome(s string) string {
	max := 0
	longest := ""
	for x := 0; x < len(s); x++ {
		even, l := center(s, x, x)
		if l > max {
			longest = even
			max = l
		}
		odd, l := center(s, x, x+1)
		if l > max {
			longest = odd
			max = l
		}
	}
	return longest
}

func center(s string, i, j int) (string, int) {
	var left, right int = i, j
	if i == j {
		left = i - 1
		right = i + 1
	}
	for x := 0; x < len(s)/2; x++ {
		if left < 0 || right >= len(s) || s[left] != s[right] {
			break
		}
		left--
		right++
	}
	return s[left+1 : right], right - left - 1
}

// @lc code=end
